[DUOS-1988][risk=no] fix zendesk support request and expand functionality to selected institutions/SOs

shaemarks opened the pull request:
Addresses [DUOS-1988](https://broadworkbench.atlassian.net/browse/DUOS-1988)
Summary of changes:
- create new `HttpClientUtils` method to send POST request without authentication header (to resolve 401 Not Authorized error)
- use `ByteArrayContent` instead of `JsonHttpContent` to send ticket (workaround to resolve 422 Unprocessable Entity error)
- expand functionality of method that creates and sends a ticket for suggested institutions and signing officials to also describe the selection or existing institutions and signing officials 
- while based on the frontend implementation of `ResearcherProfile` I would expect at most one institution  (existing or new) and one signing official (existing or new) in the `UserUpdateFields`, the endpoint itself doesn't place any restrictions on which fields appear. To accommodate this, I tried to be more flexible with the grammar of the subject and description of the support ticket 

Note: To enable support request notifications reaching Zendesk, set `activateSupportNotifications: true` under `services:` in `consent.yaml`

----
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

rushtong commented on src/main/java/org/broadinstitute/consent/http/service/SupportRequestService.java:
```diff
@@ -133,40 +141,59 @@ public void handleSuggestedUserFieldsSupportRequest(UserUpdateFields userUpdateF
         }
 
+        //Append items for ticket subject and description
+        String subject = String.format("%s user updates: %s",
+                user.getDisplayName(),
+                String.join(", ", subjectItems));
+        String description = String.format("User %s [%s] has:\n%s",
+                user.getDisplayName(),
+                user.getEmail(),
+                String.join("\n", descriptionItems));
```
There's a lot of code here in the service class that is very specific to ticket creation/population. An optional pattern would be to extract a chunk of that logic into a model class so it's more encapsulated. Not a requirement, just a style suggestion. See how we use generic utility methods on `UserUpdateFields` as a possible example.

rushtong commented on src/main/java/org/broadinstitute/consent/http/util/HttpClientUtil.java:
```diff
@@ -51,6 +51,13 @@ public HttpRequest buildPostRequest(GenericUrl genericUrl, HttpContent content,
+  public HttpRequest buildUnAuthedPostRequest(GenericUrl genericUrl, HttpContent content) throws Exception {
+    HttpTransport transport = new NetHttpTransport();
+    HttpRequest request = transport.createRequestFactory().buildPostRequest(genericUrl, content);
+    request.setHeaders(new HttpHeaders().set("X-App-ID", "DUOS"));
```
No need to change this line - just an explanation. All Terra services communicate with each other and send this header. It's intended to identify the calling app. In this usage, we're talking to Zendesk, so it will likely be ignored.

rushtong reviewed (approved):
Looks good! 👍🏽 See optional suggestions inline.

connorlbark reviewed (approved):
looks really solid! i am not sure how to test tickets going through properly, but the tests look solid and give me confidence that it works properly (sends the right payload, etc.)

shaemarks commented on src/main/java/org/broadinstitute/consent/http/models/support/SupportTicket.java:
```diff
@@ -9,10 +13,37 @@ public class SupportTicket {
+        if (Objects.isNull(name) || Objects.isNull(email)) {
+            throw new IllegalArgumentException("Name and email of user requesting support is required");
+        }
+        if (Objects.isNull(subject)) {
```
I carried these null checks over from `SupportRequestService.createSupportTicket`, but most existing model classes don't check for null values in the constructor. Thoughts on whether to keep or remove?

rushtong replied:
I think we should keep these here, they are all required fields.

shaemarks commented on src/main/java/org/broadinstitute/consent/http/service/SupportRequestService.java:
```diff
@@ -105,24 +68,25 @@ public void postTicketToSupport(SupportTicket ticket, AuthUser authUser) throws
                 try {
-                    SupportTicket ticket = createSuggestedUserFieldsTicket(userUpdateFields, user);
-                    postTicketToSupport(ticket, authUser);
+                    SupportTicket ticket = supportTicketCreator.createInstitutionSOSupportTicket(userUpdateFields, user, configuration.postSupportRequestUrl());
```
The url is included in the comment body (based on the template from [DUOS-1890](https://broadworkbench.atlassian.net/browse/DUOS-1890)). Is there a preference between passing in the url here vs having a configuration field in `SupportTicketCreator`?

rushtong replied:
If we're moving construction details to the creator class, then it does make sense to add a config there as well.

shaemarks commented:
Changes made since last approval:
- Refactor to move ticket creation/population out of `SupportRequestService` (`createSupportTicket` is now the `SupportTicket` constructor; `createInstitutionSOSupportTicket` is now located in `SupportTicketCreator`)
- Added `SupportTicketCreator` to handle creating certain types of tickets (in this case institution/SO request) since `InstitutionDAO` and `UserDAO` were needed to get names and emails for selected fields, and I didn't want the DAOs to be parameters in a SupportTicket constructor 
- moved tests into appropriate files

connorlbark reviewed (approved):
👍  changes make building a ticket much cleaner! big fan

rushtong reviewed (approved):
👍🏽 Pending whatever choices you'd like to make. Looks good!
